crates/wasi-experimental-http-wasmtime/src/lib.rs

```diff
@@ -3,190 +3,374 @@ use bytes::Bytes;
 use futures::executor::block_on;
 use http::{HeaderMap, HeaderValue};
 use reqwest::{Client, Method};
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::rc::Rc;
 use std::str::FromStr;
 use tokio::runtime::Handle;
 use url::Url;
 use wasmtime::*;
 
-const ALLOC_FN: &str = "alloc";
 const MEMORY: &str = "memory";
 
-pub fn link_http(linker: &mut Linker, allowed_hosts: Option<Vec<String>>) -> Result<(), Error> {
-    linker.func(
-        "wasi_experimental_http",
-        "req",
-        move |caller: Caller<'_>,
-              url_ptr: u32,
-              url_len: u32,
-              method_ptr: u32,
-              method_len: u32,
-              req_body: u32,
-              req_body_len: u32,
-              headers_ptr: u32,
-              headers_len: u32,
-              body_res_ptr: u32,
-              body_written_ptr: u32,
-              headers_res_ptr: u32,
-              headers_written_ptr: u32,
-              status_code_ptr: u32,
-              err_ptr: u32,
-              err_written_ptr: u32|
-              -> u32 {
-            let span = tracing::trace_span!("req");
-            let _enter = span.enter();
-            // Get the module's memory and allocation function.
-            // If either is not found, the runtime cannot write any response
-            // data, so the execution cannot continue.
-            tracing::trace!(export = MEMORY, "getting export");
-            let memory = match caller.get_export(MEMORY) {
-                Some(Extern::Memory(mem)) => mem,
-                _ => {
-                    return err(
-                        "cannot find memory".to_string(),
-                        None,
-                        None,
-                        err_ptr,
-                        err_written_ptr,
-                        1,
-                    )
+pub type WasiHandle = u32;
+
+struct Body {
+    bytes: Vec<u8>,
+    pos: usize,
+}
+
+struct Response {
+    headers: HeaderMap,
+    body: Body,
+}
+
+#[derive(Default)]
+pub struct State {
+    responses: HashMap<WasiHandle, Response>,
+    current_handle: WasiHandle,
+}
+
+#[derive(Debug, thiserror::Error)]
+enum HttpError {
```

Comment:
This is also defined in the Rust guest crate.
Is there a reason we couldn't reuse the one defined there?

(it would be similar to how we reuse the `string_to_header_map` function.)

Reply:
Hi!

Yes, this is a bit redundant.

But guest errors cannot contain any information, while host errors are Rust objects. And once we switch to WITX, guest errors will be loaded from the WITX definition, while on the host side, native Rust errors remain still have to be used.

Reply:
Ah, that makes sense.
Let's leave it like this until we move to WITX.